### DISCUSSION TITLE
Ask any questions you like!

### DISCUSSION THREAD

**Discussion by spamegg1** [spamegg1/reviews#156] [Q&A]
I'll try my best to answer. You can also ask me on Discord: `spamegg#3510`

**Comment by AhmedAzor** [comment]
why didn't you share before any information/experience about yourself in the field of 'Algorithmic' problem-solving practice? Did you try sites like Leetcode or codeforces ?  I'd like to hear from you about this, As it's mandatory for any placement in which almost a large number of your audience their main goal is to secure a job. "i don't mean how many problems you solved on leetcode or your rate on codeforces."

**Reply by spamegg1** [reply]
I've never had any experience on those sites. It never occurred to me to mention that I haven't had any experience on those sites. It's a bit strange to think of mentioning not having experience in something. It didn't come to my mind.

I think this might be a generational difference since I'm older. We learned problem solving skills in middle school and high school, not from websites. I suppose they are not teaching kids anything these days.

**Comment by clvnkhr** [comment]
Hi :) Wondering if you have looked into or done [Programming Reactive Systems](https://www.coursera.org/learn/scala-akka-reactive/home/welcome), the Akka Actors course? I have started the course but I'm worried it may not be useful enough to justify the time commitment, especially since it's still in Scala 2? (I saw there was a little about Futures in Effective Programming in Scala which I didn't do, but should be easy to quickly glance through)

**Reply by spamegg1** [reply]
Yes, I attempted that course but could not make it through (hence no review from me). It's extremely hard! I don't quite get the actor model. The lecture explanations are not as good, and the assignments are super complicated. You'll notice it has a much lower rating on Coursera than the other Scala courses. Week 1 (Futures) stuff is actually fairly easy and doable. It's the later stuff that's mind boggling. I couldn't even understand other people's solutions.

**Reply by clvnkhr** [reply]
thanks for the info! sounds like a challenge :P Based on what you said I'll try a little, but if I get super stuck I may proceed with the other courses first. If I manage to finish I may contact you again lol

**Reply by spamegg1** [reply]
Yeah, look me up on Discord / Slack / Matrix or just here. https://github.com/spamegg1/

**Comment by zhabib1997-web** [comment]
Hello ,

I heard that you have completed the OSSU curriculum, and I would like to benefit from your experience.
Is it necessary to complete all the courses included in the curriculum, or most of them?
And if this is not necessary, what are the most important courses, or the course, that one should reach in order to obtain a strong foundation in computer science?

Thank you very much

**Reply by spamegg1** [reply]
Hi :wave:  

Necessary for what exactly? :smile: I hear that it's not necessary for first jobs (which tend to be more web development oriented), but necessary to advance to higher / later positions in a career.

For strong foundation in computer science: Nand2Tetris I + II is pretty damn good all on its own. But you'll need strong programming and abstract thinking for N2T.

I'd say Core Programming + Math for CS + Nand2Tetris is a strong foundation. Then you can move in any direction you like.

But you should keep in mind that it's generally difficult to "just take \<some selection of courses\>" because there are *implicit prerequisites* that are too difficult to explain (that's why they often call it "maturity"). There are thousands of tiny little bits and pieces of required knowledge spread all over the place. There were some learners who tried to jump straight to Nand2Tetris or Algorithms, and failed badly. We generally say "just go straight in the curriculum" and "slow is the fastest way". This is the world class guitar player Dominic Miller:
<img width="681" height="593" alt="slowIsQuicker" src="https://github.com/user-attachments/assets/c02fa52a-55ca-415b-86ac-86fd525f4366" />

**Reply by zhabib1997-web** [reply]
Thank you for your quick response !!

Yes, I am talking about following the OSSU path in full and in order, where each course is a prerequisite for the next one. I am not referring to jumping ahead or taking selected courses only.

What I specifically mean is obtaining a strong foundation in programming and computer science, which would be sufficient to later specialize in areas such as web development or mobile development. From my understanding of your response, it seems that you believe completing the "Core Systems" is sufficient for this phase.

There is another point I would like to discuss:
I notice that the job market, in addition to web or mobile development, increasingly asks for some knowledge of machine learning and certain AI tools. As far as I know, these fields require a deep understanding of mathematics.

Since you are interested in mathematics, I would like to ask you:
If my goal is to have a solid mathematical foundation that enables me to study machine learning and artificial intelligence,
is it sufficient to take courses such as “Mathematics for machine learning and data science specialization” and similar ones, which focus only on the mathematics used practically in the job market?
Or does this require a deep and comprehensive foundation in mathematics, meaning revisiting and studying mathematics in general in a systematic way?

And if you believe that revisiting mathematics in a general and systematic way is necessary,
could you recommend good resources for that?

Because I notice that some popular resources such as Krista King or MathIsPower4U often focus on solving exam-style problems and what students typically face,
and do not give enough attention to the conceptual and logical depth behind mathematical ideas themselves.

Sorry for the long message.

**Reply by spamegg1** [reply]
> I am not referring to jumping ahead or taking selected courses only.

Ah OK :smile: A lot of people ask me about jumping ahead, so I reflexively assumed that :sweat_smile: 

> From my understanding of your response, it seems that you believe completing the "Core Systems" is sufficient for this phase.

Yes, but the operating systems course is extremely difficult, so I'd consider that a bit advanced. Also note Nand2Tetris, especially Part 2, requires very strong understanding of Core Programming.

>If my goal is to have a solid mathematical foundation that enables me to study machine learning and artificial intelligence, is it sufficient to take courses such as “Mathematics for machine learning and data science specialization” and similar ones, which focus only on the mathematics used practically in the job market?

Probably, but I'm not sure. I should take one of those courses and review it to be sure. The tricky part might be how quickly they go over some stuff, and what they assume from you. You need strong Linear Algebra and knowledge of Multivariable Calculus optimization such as Gradient Descent, also Probability (since a lot of neural networks and LLMs are probabilistic models). Looking at Coursera
<img width="815" height="371" alt="Captura de pantalla de 2026-01-03 20-21-43" src="https://github.com/user-attachments/assets/09a93ae1-9339-4bf1-9937-cff7284a062b" />
you can see it's similar to what I said (PCA, Principal Component Analysis, is an advanced Linear Algebra method to clean and perfect your data). If we look [at this site](https://www.math4ml.com/) it mentions the same things: Linear Algebra, Probability, calculus, optimization.

> Or does this require a deep and comprehensive foundation in mathematics, meaning revisiting and studying mathematics in general in a systematic way?

If you're going to be "mostly just a user" of Machine Learning then it's probably not necessary. If you're gonna do ML research or write your own algorithms then sure. If you struggle to understand what's going on in those "Math for ML" courses then you might need to re-visit indeed, in which case go deep on [Epp's book](https://github.com/spamegg1/Epp-Discrete-Math-5th-solutions/) then go deep in a Linear Algebra resource (either Lang's book / course or Sheldon Axler "Linear Algebra Done Right").
